File: sdk/java/src/main/java/com/infergo/Embedding.java
package com.infergo;

/**
 * Embedding model backed by ONNX session + HuggingFace tokenizer.
 *
 * <pre>{@code
 * try (Embedding embed = new Embedding("model.onnx", "tokenizer.json", "cpu", 0)) {
 *     float[] vec = embed.embed("Hello, world!");
 *     System.out.println("Dimension: " + vec.length);
 * }
 * }</pre>
 */
public class Embedding implements AutoCloseable {

    private long sessionHandle;
    private long tokenizerHandle;
    private final int maxDim;

    /**
     * Create an embedding model.
     *
     * @param modelPath     path to the ONNX embedding model
     * @param tokenizerPath path to tokenizer.json
     * @param provider      execution provider ("cpu", "cuda", etc.)
     * @param deviceId      GPU device index
     * @throws InfergoException if model or tokenizer cannot be loaded
     */
    public Embedding(String modelPath, String tokenizerPath,
                     String provider, int deviceId) {
        this(modelPath, tokenizerPath, provider, deviceId, 2048);
    }

    /**
     * Create an embedding model with explicit max dimension.
     *
     * @param modelPath     path to the ONNX embedding model
     * @param tokenizerPath path to tokenizer.json
     * @param provider      execution provider ("cpu", "cuda", etc.)
     * @param deviceId      GPU device index
     * @param maxDim        maximum embedding dimension
     * @throws InfergoException if model or tokenizer cannot be loaded
     */
    public Embedding(String modelPath, String tokenizerPath,
                     String provider, int deviceId, int maxDim) {
        this.maxDim = maxDim;

        this.sessionHandle = InfergoNative.sessionCreate(provider, deviceId);
        if (this.sessionHandle == 0) {
            String err = InfergoNative.lastErrorString();
            throw new InfergoException("Failed to create session: " + (err != null ? err : "unknown error"));
        }

        int rc = InfergoNative.sessionLoad(this.sessionHandle, modelPath);
        if (rc != 0) {
            String err = InfergoNative.lastErrorString();
            InfergoNative.sessionDestroy(this.sessionHandle);
            this.sessionHandle = 0;
            throw new InfergoException(rc, "Failed to load model: " + (err != null ? err : "unknown error"));
        }

        this.tokenizerHandle = InfergoNative.tokenizerLoad(tokenizerPath);
        if (this.tokenizerHandle == 0) {
            String err = InfergoNative.lastErrorString();
            InfergoNative.sessionDestroy(this.sessionHandle);
            this.sessionHandle = 0;
            throw new InfergoException("Failed to load tokenizer: " + (err != null ? err : "unknown error"));
        }
    }

    /**
     * Compute the embedding vector for a single text.
     *
     * @param text input text
     * @return float array of the embedding vector
     * @throws InfergoException on failure
     */
    public float[] embed(String text) {
        checkOpen();
        float[] result = InfergoNative.embedPipeline(sessionHandle, tokenizerHandle,
                                                     text, maxDim);
        if (result == null) {
            String err = InfergoNative.lastErrorString();
            throw new InfergoException("Embedding failed: " + (err != null ? err : "unknown error"));
        }
        return result;
    }

    /**
     * Compute embedding vectors for multiple texts in one call.
     *
     * @param texts array of input texts
     * @return array of embedding vectors (one per text)
     * @throws InfergoException on failure
     */
    public float[][] embedBatch(String[] texts) {
        checkOpen();
        float[] flat = InfergoNative.embedBatchPipeline(sessionHandle, tokenizerHandle,
                                                        texts, maxDim);
        if (flat == null) {
            String err = InfergoNative.lastErrorString();
            throw new InfergoException("Batch embedding failed: " + (err != null ? err : "unknown error"));
        }

        int dim = flat.length / texts.length;
        float[][] result = new float[texts.length][dim];
        for (int i = 0; i < texts.length; i++) {
            System.arraycopy(flat, i * dim, result[i], 0, dim);
        }
        return result;
    }

    /**
     * Returns the native session handle. For advanced use only.
     */
    public long getSessionHandle() {
        checkOpen();
        return sessionHandle;
    }

    /**
     * Returns the native tokenizer handle. For advanced use only.
     */
    public long getTokenizerHandle() {
        checkOpen();
        return tokenizerHandle;
    }

    @Override
    public void close() {
        if (tokenizerHandle != 0) {
            InfergoNative.tokenizerDestroy(tokenizerHandle);
            tokenizerHandle = 0;
        }
        if (sessionHandle != 0) {
            InfergoNative.sessionDestroy(sessionHandle);
            sessionHandle = 0;
        }
    }

    private void checkOpen() {
        if (sessionHandle == 0 || tokenizerHandle == 0) {
            throw new IllegalStateException("Embedding is closed");
        }
    }
}

File: sdk/dotnet/src/Infergo/Embedding.cs
namespace Infergo;

/// <summary>
/// Managed wrapper for the infergo embedding pipeline (ONNX session + tokenizer).
/// </summary>
public sealed class Embedding : IDisposable
{
    private IntPtr _session;
    private IntPtr _tokenizer;
    private bool _disposed;
    private readonly int _dim;

    /// <summary>
    /// Create an embedding pipeline.
    /// </summary>
    /// <param name="modelPath">Path to the ONNX embedding model.</param>
    /// <param name="tokenizerPath">Path to the HuggingFace tokenizer.json file.</param>
    /// <param name="dim">Embedding dimension (e.g. 384, 768, 1024).</param>
    /// <param name="provider">Execution provider: "cpu", "cuda", "tensorrt", "coreml".</param>
    /// <param name="deviceId">GPU device index (0 for first GPU).</param>
    public Embedding(string modelPath, string tokenizerPath, int dim, string provider = "cpu", int deviceId = 0)
    {
        ArgumentNullException.ThrowIfNull(modelPath);
        ArgumentNullException.ThrowIfNull(tokenizerPath);
        if (dim <= 0)
            throw new ArgumentOutOfRangeException(nameof(dim), "Dimension must be positive.");

        _dim = dim;

        _session = Native.infer_session_create(provider, deviceId);
        if (_session == IntPtr.Zero)
            throw InfergoException.FromNative($"Failed to create session with provider '{provider}'");

        int rc = Native.infer_session_load(_session, modelPath);
        if (rc != 0)
        {
            Native.infer_session_destroy(_session);
            _session = IntPtr.Zero;
            throw InfergoException.FromNative($"Failed to load embedding model: {modelPath}", rc);
        }

        _tokenizer = Native.infer_tokenizer_load(tokenizerPath);
        if (_tokenizer == IntPtr.Zero)
        {
            Native.infer_session_destroy(_session);
            _session = IntPtr.Zero;
            throw InfergoException.FromNative($"Failed to load tokenizer: {tokenizerPath}");
        }
    }

    /// <summary>
    /// Embedding dimension.
    /// </summary>
    public int Dimension => _dim;

    /// <summary>
    /// Compute the embedding vector for a text string.
    /// The full pipeline (tokenize, run ONNX, pool, normalize) executes in a single native call.
    /// </summary>
    /// <param name="text">Input text.</param>
    /// <returns>Normalized embedding vector of length <see cref="Dimension"/>.</returns>
    public float[] Embed(string text)
    {
        ObjectDisposedException.ThrowIf(_disposed, this);
        ArgumentNullException.ThrowIfNull(text);

        float[] vec = new float[_dim];
        int written = Native.infer_embed_pipeline(_session, _tokenizer, text, vec, _dim);
        if (written < 0)
            throw InfergoException.FromNative("Embedding pipeline failed");

        if (written != _dim)
        {
            // Truncate or expand to match requested dimension.
            float[] result = new float[written];
            Array.Copy(vec, result, written);
            return result;
        }

        return vec;
    }

    public void Dispose()
    {
        if (_disposed) return;
        _disposed = true;

        if (_tokenizer != IntPtr.Zero)
        {
            Native.infer_tokenizer_destroy(_tokenizer);
            _tokenizer = IntPtr.Zero;
        }

        if (_session != IntPtr.Zero)
        {
            Native.infer_session_destroy(_session);
            _session = IntPtr.Zero;
        }
    }
}
